rfbzero/experiment.py

```diff
@@ -0,0 +1,630 @@
+                 charge_first: bool = True, capacity_only: bool = True):
+        self.voltage_cutoff_charge = voltage_cutoff_charge
+        self.voltage_cutoff_discharge = voltage_cutoff_discharge
+        super().__init__(current, charge_first, capacity_only)
+
+    def run(self, duration: int, cell_model: ZeroDModel,
+            degradation: DegradationMechanism = None,
+            cls_degradation: DegradationMechanism = None,
+            ncls_degradation: DegradationMechanism = None,
+            cross_over: Crossover = None) -> CyclingProtocolResults:
+        """
+
+        Parameters
+        ----------
+        duration : int
+            Simulation time (s).
+        cell_model : ZeroDModel
+            Defined cell parameters for simulating.
+        degradation : DegradationMechanism, optional
+            Degradation mechanism(s) applied to CLS and NCLS.
+        cls_degradation : DegradationMechanism, optional
+            Degradation mechanism(s) applied to CLS.
+        ncls_degradation : DegradationMechanism, optional
+            Degradation mechanism(s) applied to NCLS.
+        cross_over : Crossover, optional
+            Crossover mechanism applied to cell.
+
+        Returns
+        -------
+        results : CyclingProtocolResults
+            Object with results from simulation
+
+        """
+
+        self._validate_model(cell_model, degradation, cls_degradation, ncls_degradation)
+
+        if degradation is not None:
+            cls_degradation = degradation
+            ncls_degradation = degradation
+
+        if not self.voltage_cutoff_discharge < cell_model.init_ocv < self.voltage_cutoff_charge:
+            raise ValueError("Ensure that 'voltage_cutoff_discharge' < 'init_ocv' < 'voltage_cutoff_charge'")
+
+        print(f"{duration} sec of cycling, time steps: {cell_model.time_increment} sec")
+
+        # initialize data results object to be sent to user
+        length_data = int(duration / cell_model.time_increment)
+        results = CyclingProtocolResults(length_data, self.capacity_only)
+
+        count = 0
+        capacity = 0.0
+
+        i_lim_cls_t, i_lim_ncls_t = cell_model.limiting_concentration(self.charge)
+
+        if self.current >= i_lim_cls_t or self.current >= i_lim_ncls_t:
+            raise ValueError("Desired current > limiting current, cell can't run")
+
+        # assign + current to charge, - current to discharge
+        i = self.current_direction() * self.current
+
+        losses, _, _ = cell_model.total_overpotential(i, self.charge, i_lim_cls_t, i_lim_ncls_t)
+        ocv = cell_model.open_circuit_voltage()
+        cell_v = cell_model.cell_voltage(ocv, losses, self.charge)
+
+        if cell_v <= self.voltage_cutoff_discharge or cell_v >= self.voltage_cutoff_charge:
+            raise ValueError("Desired current too high, overpotentials place cell voltage outside voltage cutoffs")
+
+        while count != length_data:
```

Comment:
as discussed, can't use for loop due to iterations `!=` required number of timesteps. occurs because cycles often require an extra step, to determine that voltage/current/capacity limits have been exceeded